core/templates/components/common-layout-directives/navigation-bars/top-navigation-bar.component.ts

```diff
@@ -398,7 +401,7 @@ export class TopNavigationBarComponent implements OnInit, OnDestroy {
     }
 
     // If i18n hasn't completed, retry after 100ms.
-    if (!this.checkIfI18NCompleted) {
+    if (this.I18nCompleteTest || !this.checkIfI18NCompleted) {
```

Comment:
Not sure if I understand why was this added.

Reply:
In one of the tests for `truncateNavbar`, when it calls `checkIfI18NCompleted` in the if condition, its forced to return false regardless of the actual return value, so we may test if `truncateNavbar` retries after 100ms.

The original piece of code simply assigned null to `checkIfI18NCompleted` and called `truncateNavbar`, to force it to retry the operation.

After enabling strict checks, it was one of the errors, since assigning `null` to a method of type `() => boolean` is not allowed.

This was my attempt at finding a workaround.

--In normal circumstances, `I18nCompleteTest` is false, so the if condition is evaluated solely on the basis of the return value of `checkIfI18NCompleted`.
--During the test, `I18nCompleteTest` is set to true (line 344 of `top-navigation-bar.component.spec.ts`), so it simulates a situation where `checkIfI18nCompleted` returns false (the if condition is evaluated as if it returns false), and `truncateNavbar` is forced to retry after 100ms.

Reply:
I think that a better solution for this would be to assign `false` to `checkIfI18NCompleted` to verify if the test retries.

Reply:
But `checkIfI18NCompleted` is a method and assigning a value to it causes this type-mismatch error:
![image](https://user-images.githubusercontent.com/78612244/138133270-8c5a6ea1-5bcc-4eef-887a-04aadde3b232.png)

Is this what you suggested, or have I misunderstood you maybe?

Reply:
You can assign `(): boolean => false;` to it.

Reply:
Tried this. This seemed obvious, however the test still fails and even though the method now returns false, `truncateNavbar` doesn't retry:
![image](https://user-images.githubusercontent.com/78612244/138551011-d08a8811-6a26-4340-ac42-e519de71bb67.png)

and continues to execute the rest of the function (like calling `document.querySelector(div.collapse.navbar-collapse)` in the next step).

The test expects the rest of the function to NOT be executed, therefore it fails.
The other option is to assign null to the method (like it was originally), but that causes typescript errors.

I'm not sure what's causing it, or how to proceed here.

Reply:
Wait, actually I think I see the problem now. the `checkIfI18NCompleted` is a function, but here we do not call it as a function. So here it should be 
```suggestion
    if (!this.checkIfI18NCompleted()) {
```
and all other places should be modified accordingly.

Reply:
That was the one of the first things I noticed too. However, for some reason, adding a pair of parentheses at the end, and calling it like a function gives this error:
![image](https://user-images.githubusercontent.com/78612244/138560577-be8ae53a-fe6a-4a3f-9146-d2336113cd27.png)

The second line is really confusing because it obviously IS a function, or a method atleast.

Reply:
Can you please push the changes, where this error is happening, so I can take a look at your current state of the code?

Reply:
I'm not sure if I understand this:

> push the changes, where this error is happening

I have pushed the most recent changes, and the code in this PR represents the current state of the code on my machine, if that's what you mean.
@vojtechjelinek

Reply:
I mean the chages where you get the frontend error you mention above.

Reply:
The method definitions for `checkIfI18NCompleted` and `truncateNavbar` originally looked like this:
![image](https://user-images.githubusercontent.com/78612244/138629794-d564ae6c-61d3-452e-b7a3-d5499c791fc4.png)

The unit test originally looked like this:
![image](https://user-images.githubusercontent.com/78612244/138629865-e8bc306f-ab43-49af-a028-6435e97d8869.png)

_This was the state of the code before I picked up this issue._

But, when I make the following changes
-to the unit test:
![image](https://user-images.githubusercontent.com/78612244/138629989-4d53a4f0-18ef-44b5-bf27-d64559385dd4.png)

-and to the method definition:
![image](https://user-images.githubusercontent.com/78612244/138630091-733a4bac-3cf1-4b99-a6dd-077cebec049c.png)

this is when I get the frontend error i mentioned above:
![image](https://user-images.githubusercontent.com/78612244/138630227-acdc93b8-c96e-4591-ae2c-fa0d8d82df28.png)

Reply:
@vojtechjelinek

Reply:
Hmmm, this is really weird, I tried using `spyOn` and it throws the same error. @aishwary023 @kevintab95 @nithusha21 any ideas how to handle this?

Reply:
Do you mean you tried to `spyOn(component, "checkIfI18NCompleted").and.returnValue(false)`?

Reply:
Is this.checkIfI18NCompleted rewritten anywhere else? It seems to be rewritten in the original code, but are there some other places where it gets rewritten which could cause an issue?

Reply:
I don't think so.

In the top-navigation-bar.component.ts file, it appears twice: once when it is defined, once in the If statement in the `truncateNavbar` method.

In the top-navigation-bar.component.spec.ts file, it appears 5 times:
-four times in `expect` and `spyOn` statements:
![image](https://user-images.githubusercontent.com/78612244/139093399-c041075c-ea34-4b38-9782-35ecdac1ae6f.png)

And once where the function is redefined to return false (in the test we're talking about).

Reply:
@nithusha21 
Have you been able to figure out what might be causing the issue or how to fix it?
I've tried the solutions suggested in this thread, and some other things (like changing the `checkIfI18NCompleted` method from a method, to a property of type function) and I'm not sure how else to proceed.

I would appreciate it if you could give me some pointers on how I may fix this.

Reply:
@nithinrdy Hey, can you try to find minimal code that still produces the error? That should help us with analyzing this issue.

Reply:
I've tried stripping away as much code as I can-

The 'top-navigation-bar.component.ts' file now **contains**:
-->the import statements
-->the component decorator
-->In the TopNavigationBarComponent class, this is what is left:
1. The constructor
2. The methods that are causing this problem, `checkIfI18NCompleted` and **_part of_** `truncateNavbar` (I've commented out some parts of the method):
![image](https://user-images.githubusercontent.com/78612244/139710702-3332b857-2c02-451e-994e-db5aa2dcfd1e.png)

3. and the `ngOnDestroy()` method.

-->And finally, the statement that registers a new directive at the end of the file:
![image](https://user-images.githubusercontent.com/78612244/139711031-32bbb251-a1d4-4704-b509-e96a9fa88f38.png)

_Other than things listed, everything in 'top-navigation-bar.component.ts' has been commented out (the entire ngOnInit() hook, all the class properties, all the class methods, etc)._

In the 'top-navigation-bar.component.spec.ts' file, I've **_commented out every spec test, except the one we are trying to fix_** (this one):
![image](https://user-images.githubusercontent.com/78612244/139713136-e91af96b-0b27-4ff4-b7e7-cfded97bc356.png)

Everything else (the `beforeEach` functions, the import statements, etc.) has been left completely intact.

At this stage, I still get the following error:
![image](https://user-images.githubusercontent.com/78612244/139712995-068d87f8-b6c2-434e-aac4-e7e77064176d.png)

Also, I'm not sure if this is relevant, but grepping the root directory for a list of files containing the string `checkIfI18NCompleted` gives this result:
![image](https://user-images.githubusercontent.com/78612244/139712503-2491f91a-e962-4369-80f9-cef7a4f87000.png)

Reply:
@vojtechjelinek PTAL and let me know if this is what you were looking for

Reply:
Hmm, interesting, one more thing can you please prepare a [debugging doc](https://github.com/oppia/oppia/wiki/Debugging-Docs) for this? I will then share it with other folks that might be able to help. Also, you can probably ask about this error on StackOverflow, now that you have a minimal example.

For the grepped files, I think it is correct, the cache files should pose a problem and the webpack bundle is regenerated every time the tests start.

Reply:
@vojtechjelinek
I've created [the doc](https://docs.google.com/document/d/1LzwNPu036SmHPCd2WEuKEL2-8e2efDGvc3GwwzTyi0M/edit?usp=sharing). This is the first time I've created such a document. Please let me know if it needs any changes, or additions.

I will be creating a stackoverflow thread soon, since I am currently bogged down by assignments.

Reply:
Hey, sorry for slight delay. Sean figured it out 
![image](https://user-images.githubusercontent.com/10142938/140733391-68776bd3-2c02-4a6a-a010-7a38ef6ff74c.png)

Reply:
@vojtechjelinek I have made the changes, but have yet to push them. I think a small comment (right above the `let that = this;` assignment) would be a good idea. Something like:

> Inside a setTimeout function call, 'this' points to the global object. To access the context in which the setTimeout call is made, we need to first save a reference to that context in a variable, and then use that variable in place of the 'this' keyword.

Let me know what you think of this. If it is unnecessary, I will push the changes without it.

Reply:
Yep, comment would be nice.

Reply:
Done.

Reply:
The E2E tests have all failed. I don't know the cause, but the error `this.checkIfI18NCompleted is not a function` occurred multiple times in each of the ten tests. How should I proceed here?

P.S. The error originates from the `/webpack_bundles/` directory. If I'm not mistaken, I remember you stating that the files in this directory are regenerated during the tests? I'm not sure what this means. PTAL

Reply:
This is a wild guess, but I'm assuming the error is caused by the fact that `this.checkIfI18NCompleted` is called in the setTimeoutcall (when `truncateNavbar` calls itself) and since at that point `this` refers to the global object, the error occurs.

Reply:
@seanlip Hey, it seems that even after your fix the e2e tests fail with the same error as the frontend tests before (`this.checkIfI18NCompleted is not a function`), any idea why that can still be happening.

@nithinrdy Meanwhile you can try debugging this yourself, you can run the e2e tests locally and add some debugging logging statements.

Reply:
I don't know, sorry, and I'm afraid I don't have the bandwidth at the moment to debug. You will probably need to dig and do some logging.

Reply:
One question though -- do the tests run fine locally? Wondering if the problem is the tests itself, or the local vs CI environment -- that might give a clue.

Reply:
@seanlip I ran one of the e2e tests locally, and it does fail with the same 'not a function' error. Although I have not run every single test, I think it's safe to assume the rest would fail too.

I've been looking into it but no luck so far. Will see if I can make it work.

Reply:
Do the frontend tests run fine locally?

When you start the server locally, and play around with it, do you get console errors? If you do, that provides a nice debugging environment. (Try to simulate the behaviour of the e2e tests manually, perhaps.)

Reply:
I think I've managed to fix it. There were some more instances of `truncateNavbar` being called (in the `ngOnInit()` hook) in a setTimeout call, using `this` (which weren't picked up by the frontend tests for some reason even though the `ngOnInit()` call _is_ made in some of the unit tests?).

I've made some changes and, so far:

1. The console on my local server looks clean
2. The frontend tests all pass (although they passed earlier too.)
3. More importantly, the one e2e test that I ran locally which gave me the 'not a function' error, now passes just fine.

Although one passed, I want to run all the e2e tests before I push the changes, just in case. I'm a little busy atm so I will have to run them sometime later. As it stands, I think we've fixed it.

Reply:
There were some more instances of `truncateNavbar` being called inside `setTimeout` calls using `'this'` (inside the `ngOnInit()` hook) which, for some reason, were not picked up by the frontend tests (even though `ngOnInit()` is called multiple times in the unit tests).

I've modified them and the e2e tests are now passing.

Reply:
@seanlip Although we've fixed it, I'm still a little confused about the problem I've encountered. I'd appreciate it if you could clarify this for me.

You claimed that the reason `this.checkIfI18NCompleted` was not being recognized as a function is because, inside a `setTimeout` call, `this` loses its meaning. I looked into it and sure enough, inside setTimeout calls `this` points to the global object instead of the current object. So it makes sense that `checkIfI18NCompleted` wasn't found on the global object because it is associated with the `TopNavigationBarComponent` class.

But, by the same token, shouldn't the error have been '`this.truncateNavbar` is not a function'? Since it is the first thing to be called in the setTimeout call and it isn't associated with the global object either.

Similarly, in the `truncateNavbar` method definition, `this.checkIfI18NCompleted` is the second function to be called. The first is `this.windowDimensionsService?.isWindowNarrow()`, which, again, should have caused a similar error.

Do you happen to know why `this.checkIfI18NCompleted` was the only one to throw this error?

Reply:
It's a good question, and offhand, I do not know. I didn't look into it deeply enough, but unfortunately I don't have the bandwidth to do so now. You can probably get more insight by doing some logging of what `this.truncateNavbar` was (and also what `this` pointed to in that context), I think.

One possible thought for your last point though -- is that maybe `this.windowDimensionsService` was undefined and so `this.windowDimensionsService?.isWindowNarrow()` might not have done anything? I'm not sure, but again you should be able to get some insight by logging.